### PR TITLE
52 battle flow

### LINKED ISSUE #52: After battle, units should listen for the ExitBattleMessage from battle manager and deal / take damage accordingly

### PR DESCRIPTION
**Pull request by armsnyder**
closes #52

### CODE REVIEW

**Comment by armsnyder** [comment]
As I mentioned in class today, this branch contains both the new battle mechanics and fixes for many bugs in our code that were previously undetectable because we didn't have the full game flow put together yet. That, and we don't unit test the code.

The code is fully commented and should be readable. Some of the bug fixes are legitimate, but others are hacky. I marked all the hacky ones with TODOs.

Tomorrow I'll take some time to look through the code myself, add in-line GitHub comments where necessary, and create mini-issues based around turning some of my hacky fixes into real, scalable fixes.

File: Assets/scripts/BattleManager.cs [hunk 1]
```diff
@@ -126,9 +196,18 @@ private class PlayerBattleData {
 			// TODO: Add support for HOPOs
 			bool noteWasHit = false;
 			// Go through the possible notes that the player could have been trying to hit
+			GameObject[] noteObjects = GameObject.FindGameObjectsWithTag("noteObject");
 			foreach (Note n in hitNotes) { // Warning! O(n^2)
-				if (noteWasHit)
+				if (noteWasHit) {
```
Comment (armsnyder) [inline Assets/scripts/BattleManager.cs:201]: @GregoryElliott Take a look at this section of code for generating visual feedback when a note is hit. Currently the code attempts to destroy the hit note, but it clearly isn't working.

File: Assets/scripts/UnitManager.cs [hunk 2]
```diff
@@ -186,4 +188,20 @@ void Start()
 		}
 
 	}
+
+	void OnExitBattle(ExitBattleMessage m) {
```
Comment (armsnyder) [inline Assets/scripts/UnitManager.cs:192]: @GregoryElliott Look at this chunk of code for why damage isn't being dealt. Try adding a breakpoint here and see what you get.

File: Assets/scripts/MelodyUnit.cs [hunk 3]
```diff
@@ -103,4 +103,22 @@ public override void UnMark()
     {
         GetComponent<Renderer>().material.color = LeadingColor;
     }
+
+	public void DealDamage(MelodyUnit other, float damagePercent)
```
Comment (armsnyder) [inline Assets/scripts/MelodyUnit.cs:107]: @GregoryElliott This could be another place to look for why damage isn't working.

**Comment by cegrief** [comment]
THIS IS MAHVELOUS

**Pull request merged**